File: 0237-delete-node-in-a-linked-list/inplace-node-deletion/cpp/solution.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    void deleteNode(ListNode* node) {
        ListNode* temp = node->next;   // Preserve pointer to next node
        node->val = temp->val;         // Copy next node's value
        node->next = temp->next;       // Skip next node
        delete temp;                   // Free memory
    }
};

File: 0237-delete-node-in-a-linked-list/inplace-node-deletion/java/solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public void deleteNode(ListNode node) {
        node.val = node.next.val;      // Copy next node's value
        node.next = node.next.next;    // Skip next node
        // Java's garbage collector automatically frees memory of the skipped node
    }
}
